File: app/routers/academy.py
"""Guita Coach Academy — contenido educativo fijo, priorizado por perfil del usuario."""
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session
from ..database import get_db
from ..models import User, Investment
from ..security import get_current_user
from ..services.academy_content import TOPICS, CATEGORY_LABELS, GLOSSARY

router = APIRouter(prefix="/api/academy", tags=["academy"])


def _build_profile(user: User, investments: list[Investment]) -> dict:
    asset_types = {inv.asset_type for inv in investments if inv.asset_type}
    return {
        "is_beginner": len(investments) == 0,
        "crypto_only": bool(asset_types) and asset_types == {"crypto"},
        "low_buffer": (user.ahorro_pct or 0) < 10,
        "variable_income": bool(getattr(user, "income_is_variable", False)),
        "diversified": len(asset_types) >= 2,
    }


def _score(topic: dict, profile: dict) -> int:
    return sum(1 for tag in topic["tags"] if profile.get(tag))


@router.get("")
def get_academy(db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    investments = db.query(Investment).filter(
        Investment.user_id == user.id,
        Investment.status == "open",
    ).all()
    profile = _build_profile(user, investments)

    scored = [{**t, "score": _score(t, profile)} for t in TOPICS]
    recommended = sorted([t for t in scored if t["score"] > 0], key=lambda t: -t["score"])[:4]
    if not recommended:
        recommended = [t for t in scored if t["category"] == "primeros_pasos"]

    by_category: dict = {}
    for t in scored:
        by_category.setdefault(t["category"], []).append(t)

    categories = [
        {"category": cat, "label": CATEGORY_LABELS.get(cat, cat), "topics": topics}
        for cat, topics in by_category.items()
    ]

    return {
        "recommended": [{k: v for k, v in t.items() if k != "score"} for t in recommended],
        "categories": categories,
        "glossary": GLOSSARY,
    }

File: services/ai-service/src/routes/academy.js
/** Guita Coach Academy — contenido educativo fijo, priorizado por perfil del usuario. */
import { Router } from "express";
import { prisma } from "@guita-coach/db";
import { requireAuth, ah } from "@guita-coach/shared";
import { TOPICS, CATEGORY_LABELS, GLOSSARY } from "../services/academyContent.js";
export const academyRouter = Router();
academyRouter.use(requireAuth);
function buildProfile(user, investments) {
  const assetTypes = new Set(investments.map(i => i.assetType).filter(Boolean));
  return {
    is_beginner: investments.length === 0,
    crypto_only: assetTypes.size > 0 && assetTypes.size === 1 && assetTypes.has("crypto"),
    low_buffer: (user.ahorroPct || 0) < 10,
    variable_income: Boolean(user.incomeIsVariable),
    diversified: assetTypes.size >= 2
  };
}
function score(topic, profile) {
  return topic.tags.reduce((s, tag) => s + (profile[tag] ? 1 : 0), 0);
}
academyRouter.get("/", ah(async (req, res) => {
  const user = req.user;
  const investments = await prisma.investment.findMany({
    where: {
      userId: user.id,
      status: "open"
    },
    select: {
      assetType: true
    }
  });
  const profile = buildProfile(user, investments);
  const scored = TOPICS.map(t => ({
    ...t,
    score: score(t, profile)
  }));
  let recommended = scored.filter(t => t.score > 0).sort((a, b) => b.score - a.score).slice(0, 4);
  if (!recommended.length) recommended = scored.filter(t => t.category === "primeros_pasos");
  const byCategory = new Map();
  for (const t of scored) {
    if (!byCategory.has(t.category)) byCategory.set(t.category, []);
    byCategory.get(t.category).push(t);
  }
  const categories = [...byCategory.entries()].map(([cat, topics]) => ({
    category: cat,
    label: CATEGORY_LABELS[cat] || cat,
    topics
  }));
  res.json({
    recommended: recommended.map(({
      score: _s,
      ...rest
    }) => rest),
    categories,
    glossary: GLOSSARY
  });
}));
